clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -169,6 +174,19 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
             return NetworkClientDelegate.PollResult.EMPTY;
         }
 
+        pollTimer.update(currentTimeMs);
+        if (!isLeaveGroupInProgress() && pollTimer.isExpired()) {
+            logger.warn("consumer poll timeout has expired. This means the time between subsequent calls to poll() " +
+                "was longer than the configured max.poll.interval.ms, which typically implies that " +
+                "the poll loop is spending too much time processing messages. You can address this " +
+                "either by increasing max.poll.interval.ms or by reducing the maximum size of batches " +
+                "returned in poll() with max.poll.records.");
+            membershipManager.leaveGroup().whenComplete((ignored, exception) -> {
```

Comment:
What is the exact to do in KIP-848 if the poll timer expires? The current implementation sends a leavegroup request directly without revoking partitions.  If we want to do that, we need to modify the state transition a bit.